### DISCUSSION TITLE
Ideas for projects in ableC

### DISCUSSION THREAD

**Discussion by remexre** [melt-umn/silver#596] [Ideas]
It might be a good thing to have a realish piece of software written in ableC, as a way to show off ableC, a way to find the "rough spots" in using it, and as a way to find new extensions to write.

I'd propose an ircd for this -- IRC's got simple enough semantics (chat with channels and DMs, no persistent accounts or history or anything) that it should be possible for a student to write one in a month or two in plain C, plus it has lots of places where ableC would be useful:

- Something goroutine-like (could be mcontext_t tasks + channels) would be very useful for concurrency; a simple model would be:
  - one task per channel
  - two tasks per user, one for sending and one for recving
  - two (go-like) channels per user per channel, acting as a single bidirectional one
  - user send task reads the channel messages and sends them over the socket
  - user recv task reads+parses from the socket and sends to the appropriate channel
- The IRC protocol is weird, so a generated parser would be useful!
- The IRC protocol has a small stateful structure, so maybe an extension for state machines would be useful?
- String manipulation and some data structures are involved, so:
  - more readable code!
  - faster code, if we monomorphize+inline (see also, well-written c++/rust are faster than well-written c)
  - being able to say "fewer memory unsafety bugs b/c ableC-{string,vector,map} make making mistakes harder"
- ircds are typically taking untrusted input from the net, so they're a "real example" as far as security-related analyses are concerned.
- I want a new ircd, ngircd's kinda crappy

**Comment by ericvanwyk** [comment]
This would be good.  Any pointers to IRC protocol docs or intro to using IRC? Something that would give someone a place to get started thinking about this.

**Reply by remexre** [reply]
For an intro to using IRC, https://libera.chat/guides/basics

I'd probably recommend [the Libera webchats] or [Hexchat] to someone wanting to try it out; we have #silver on Libera bridged to the Slack channel, so that'd be a reasonable place to try posting a "hello from irc" or whatever. (My current "main" setup is based on [Weechat]+[soju], but Weechat is "power-user-ware...")

For a brutally simple client to study, [ii] comes in at just under 1k lines of C. I've got a parser for the lexical syntax (in Haskell, using parser combinators) [here][Network.LemonIRC.Client.RawMessage], though that might not be the *most* useful. Pretty much every popular programming language has an IRC client library available too. (I'd still recommend reading [RFC2810] and [RFC2811], they're 29 pages together and describe the semantics pretty well.)

The "It's very simple — you read the protocol and write the code" approach would probably start at RFCs [2810][RFC2810], [2811][RFC2811], and [2812][RFC2812] (the first two are semantics, last is protocol), [RFC7194] (just specifies how TLS should work, the answer is "at port 6697" but this requires 3 pages to say), and then consider going on to the [IRCv3] specs, which add all the stuff modern chat protocols have, eg typing notifs, replies, reacts, server-side history, etc. All the IRCv3 stuff is optional, clients negotiate which features are enabled with the server, and I'd say probably skip implementing them unless one demonstrates a particular extension we want to show off (e.g. sqlite ext for server-side history).

Depending on the client, it may be possible to see what is being sent and received too; in [Weechat] (not to be confused with Wechat!) this is `/server raw`:

![](https://cdn.remexre.xyz/files/b2d3992328c17b5627d083a416827c13ad100bab.png)

[RFC2810]: https://datatracker.ietf.org/doc/html/rfc2810
[RFC2811]: https://datatracker.ietf.org/doc/html/rfc2811
[RFC2812]: https://datatracker.ietf.org/doc/html/rfc2812
[RFC7194]: https://datatracker.ietf.org/doc/html/rfc7194
[IRCv3]: https://ircv3.net/irc/
[Hexchat]: https://hexchat.github.io/
[Weechat]: https://weechat.org/
[ii]: https://git.suckless.org/ii/files.html
[the Libera webchats]: https://libera.chat/guides/webchat
[soju]: https://soju.im/
[Network.LemonIRC.Client.RawMessage]: https://git.sr.ht/~remexre/lemonbot/tree/b1ef0ccb6de67bd737820f0b25314eccda950171/item/src/Network/LemonIRC/Client/RawMessage.hs
